src/MonitorProfiler/MainProfiler/ExceptionTracker.cpp

```diff
@@ -0,0 +1,301 @@
+
+    return S_OK;
+}
+
+HRESULT ExceptionTracker::ExceptionThrownStackSnapshotCallback(
+    FunctionID funcId,
+    UINT_PTR ip,
+    COR_PRF_FRAME_INFO frameInfo,
+    ULONG32 contextSize,
+    BYTE context[],
+    void* clientData)
+{
+    // Callback is only used to get the function ID of the top frame.
+    *static_cast<FunctionID*>(clientData) = funcId;
+
+    // Cancel stack snapshot callbacks after the top frame.
+    return S_FALSE;
+}
+
+HRESULT ExceptionTracker::GetFullyQualifiedMethodName(FunctionID functionId, tstring& fullMethodName)
+{
+    HRESULT hr = S_OK;
+
+    ClassID classId;
+    ModuleID moduleId;
+    mdToken token;
+    IfFailLogRet(m_pCorProfilerInfo->GetFunctionInfo(
+        functionId,
+        &classId,
+        &moduleId,
+        &token
+    ));
+
+    if (0 == (token & mdtMethodDef))
+    {
+        return E_FAIL;
+    }
+    mdMethodDef methodDef = token;
+
+    ComPtr<IMetaDataImport2> pMetadataImport;
+    IfFailLogRet(m_pCorProfilerInfo->GetModuleMetaData(
+        moduleId,
+        CorOpenFlags::ofRead,
+        IID_IMetaDataImport2,
+        (IUnknown**)&pMetadataImport
+    ));
+
+    // Get Module Name
+    ULONG count = 0;
+    IfFailLogRet(m_pCorProfilerInfo->GetModuleInfo(
+        moduleId,
+        nullptr,
+        0,
+        &count,
+        nullptr,
+        nullptr
+    ));
+
+    unique_ptr<WCHAR[]> pModulePath(new (nothrow) WCHAR[count + 1]);
+    IfFailLogRet(m_pCorProfilerInfo->GetModuleInfo(
+        moduleId,
+        nullptr,
+        count + 1,
+        nullptr,
+        pModulePath.get(),
+        nullptr
+    ));
+
+    // Get Class Name
+    tstring className;
+    if (0 == classId)
+    {
+        // Probably a value type e.g. struct
+        className.assign(_T("[Unknown]"));
+    }
+    else
+    {
+        mdTypeDef typeDef = mdTokenNil;
+        IfFailLogRet(m_pCorProfilerInfo->GetClassIDInfo(
+            classId,
+            nullptr,
+            &typeDef
+        ));
+
+        IfFailLogRet(pMetadataImport->GetTypeDefProps(
+            typeDef,
+            nullptr,
+            0,
+            &count,
+            nullptr,
+            nullptr
+        ));
+
+        unique_ptr<WCHAR[]> pClassName(new (nothrow) WCHAR[count + 1]);
+        IfFailLogRet(pMetadataImport->GetTypeDefProps(
+            typeDef,
+            pClassName.get(),
+            count + 1,
+            nullptr,
+            nullptr,
+            nullptr
+        ));
+
+        className.assign(pClassName.get());
+    }
+
+    // Get Method Name
+    IfFailLogRet(pMetadataImport->GetMethodProps(
+        methodDef,
+        nullptr,
+        nullptr,
+        0,
+        &count,
+        nullptr,
+        nullptr,
+        nullptr,
+        nullptr,
+        nullptr
+    ));
+
+    unique_ptr<WCHAR[]> pMethodName(new (nothrow) WCHAR[count + 1]);
+    IfFailLogRet(pMetadataImport->GetMethodProps(
+        methodDef,
+        nullptr,
+        pMethodName.get(),
+        count + 1,
```

Comment:
Feel free to ignore this (and any other comments) since I'm not used to reading through these types of patterns - so, I believe I understand how these three 'get' sections are working after looking through some docs, but I realized we keep reusing the same `count` variable for all of them. It appears that we only read from it directly after getting its updated value for the length of a module/class/method name, but it did introduce some confusion as to what the responsibility of `count` was since it's being used for unrelated(ish) purposes - depending where you look in the file, it can _really_ mean `moduleNameLength`, `classNameLength`, or `methodNameLength` (I think). Just an idea for readability, unless there's a reason not to do it.

Reply:
Yes, these could be separate locals, but they are not used outside of the scope of allocating the array for each type of information we get. I don't think there is any actual noticeable impact if there is one vs three of these. I don't feel strongly either way, but if it makes it easier to read, then I can separate them.